### PR TITLE
[0.60] Ensure WebSocket write lifetime

### PR DESCRIPTION
**Pull request by JunielKatarn**
Backport of #4790

Resume point in `WinRTWebSocketResource::PerformWrite` is represented as a functor when awaited by the member `Mso::DispatchQueue`.

Such functor is currently implicitly created within `resume_on_queue::await_suspend` in the stack.

This change makes the functor a member variable so no stack allocations happen within `await_suspend`.

See https://devblogs.microsoft.com/oldnewthing/20191209-00/?p=103195

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4818)

### CODE REVIEW

**Comment by NickGerleman** [comment]
You'll also want to get 0.61 (and mark the change type as patch instead of prerelease for that branch).

I can get this for 0.62, but usually we mark "backport to xxx" on the original PR so we pick things up in bulk.

**Comment by JunielKatarn** [comment]
> You'll also want to get 0.61 (and mark the change type as patch instead of prerelease for that branch).
> 
> I can get this for 0.62, but usually we mark "backport to xxx" on the original PR so we pick things up in bulk for the most recent release.

Noted.
I'll follow that convention going forward.
Note, while I'm porting to 0.60 first, I plan to port any master changes to every stable branch for the Networking components.

**Comment by NickGerleman** [comment]
Just saw some changes to this branch that weren't in master. We really shouldn't be doing that or we will see regressions during upgrades.

**Comment by JunielKatarn** [comment]
> Just saw some changes to this branch that weren't in master. We really shouldn't be doing that or we will see regressions during upgrades.

I'm addressing all of those today. Please unblock this PR.

**Comment by NickGerleman** [comment]
@junielkatarn please follow the standard procedure of getting a review for changes snd backporting to all the branches. We can't have different changes in different branches, especially if unreviewed and in a cherry-pick CT. It's too hard to track.

**Comment by JunielKatarn** [comment]
> @JunielKatarn please follow the standard procedure of getting a review for changes snd backporting to all the branches. We can't have different changes in different branches, especially if unreviewed and in a cherry-pick CR. It's too hard to track.

I would appreciate being directed to the standard's document or formal source.
Else, it's a common-sense convention, and not a standard.

As mentioned before, I'm syncing the changes in all branches today.
It's not trivial to send up to 4 PRs in descending order to get a patch required mainly for the oldest branch (`0.60-stable`).

I'll send the PRs shortly to satisfy this convention.

**Comment by NickGerleman** [comment]
We're working to get Office to 0.62 to make the situation better. The changes should be straight cherry-picks though. Please revert any changes that aren't what you've already checked in.

**Comment by JunielKatarn** [comment]
@NickGerleman 
Backport PRs for all other stable branches have been issued.

Please unblock.

**Comment by JunielKatarn** [comment]
We should develop a backport automation strategy.
It is not impossible that we will end up with several stable branches to maintain.

**Comment by NickGerleman** [comment]
@junielkatarn yes, this state isn't great right now. I've been doing bulk cherry picks into 0.62 for now. Ideally we should only have one stable branch to maintain at a time, but the decision for Office to skip 0.61, and time sensitivity of some of your changes has made things trickier.

It does end up being simpler even from a dev perspective if we do straight cherry-picks of what was squashed, since it's easy to apply a change directly to a different branch.

File: vnext/Desktop/Modules/WebSocketModule.cpp [hunk 1]
```diff
@@ -24,7 +25,15 @@ constexpr char moduleName[] = "WebSocketModule";
 
 namespace Microsoft::React {
 
-WebSocketModule::WebSocketModule() {}
+WebSocketModule::WebSocketModule()
+    : m_resourceFactory{[](const string &url, bool legacyImplementation, bool acceptSelfSigned) {
+        return IWebSocketResource::Make(url, legacyImplementation, acceptSelfSigned);
+      }} {}
```
Comment (NikoAri) [inline vnext/Desktop/Modules/WebSocketModule.cpp:31]: (nit) parameter independent settings like this should be specified in .h file.
For constructor initialization, it's better to only specify data that comes with constructor's arguments

File: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp [hunk 2]
```diff
@@ -64,14 +66,16 @@ namespace
 
       void await_suspend(std::experimental::coroutine_handle<> resume)
       {
-        m_queue.Post([context = resume.address()]() noexcept
+        m_callback = [context = resume.address()]() noexcept
```
Comment (NikoAri) [inline vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:69]: >m_callback [](start = 8, length = 10)

Wasn't this already part of another change?

**Comment by NikoAri** [comment]
>     m_socket.Control().IgnorableServerCertificateErrors().Append(certException);

(nit) we should do std::move here.

---
Refers to: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:97 in b79bbb1. [](commit_id = b79bbb1bdd76d1236327536c7287ea490f61af3f, deletion_comment = False)

File: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp [hunk 3]
```diff
@@ -91,8 +98,13 @@ WinRTWebSocketResource::WinRTWebSocketResource(Uri&& uri, vector<ChainValidation
   }
 }
 
-WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString, vector<ChainValidationResult> certExeptions)
-  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) }, certExeptions)
+WinRTWebSocketResource::WinRTWebSocketResource(IMessageWebSocket&& socket, Uri&& uri, vector<ChainValidationResult> certExeptions)
```
Comment (NikoAri) [inline vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:101]: >> ce [](start = 114, length = 4)

&&

File: vnext/ReactWindowsCore/WinRTWebSocketResource.h [hunk 4]
```diff
@@ -20,9 +20,10 @@ namespace Microsoft::React {
 
 class WinRTWebSocketResource : public IWebSocketResource, public std::enable_shared_from_this<WinRTWebSocketResource> {
   winrt::Windows::Foundation::Uri m_uri;
-  winrt::Windows::Networking::Sockets::MessageWebSocket m_socket;
-  winrt::Windows::Networking::Sockets::MessageWebSocket::MessageReceived_revoker m_revoker;
-  winrt::Windows::Storage::Streams::DataWriter m_writer{m_socket.OutputStream()};
+  winrt::Windows::Networking::Sockets::IMessageWebSocket m_socket;
+  // TODO: Use or remove.
```
Comment (NikoAri) [inline vnext/ReactWindowsCore/WinRTWebSocketResource.h:24]: > // TODO: Use or remove. [](start = 1, length = 24)

Yes :)

File: vnext/ReactWindowsCore/WinRTWebSocketResource.h [hunk 5]
```diff
@@ -44,6 +45,7 @@ class WinRTWebSocketResource : public IWebSocketResource, public std::enable_sha
   std::function<void(Error &&)> m_errorHandler;
 
   WinRTWebSocketResource(
+      winrt::Windows::Networking::Sockets::IMessageWebSocket &&socket,
       winrt::Windows::Foundation::Uri &&uri,
       std::vector<winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult> certExeptions);
```
Comment (NikoAri) [inline vnext/ReactWindowsCore/WinRTWebSocketResource.h:50]: >> c [](start = 93, length = 3)

&&

File: vnext/ReactWindowsCore/WinRTWebSocketResource.cpp (hunk 2 above)
Comment (JunielKatarn) [reply vnext/ReactWindowsCore/WinRTWebSocketResource.cpp:69]: The other change was merged against master. This one want against `0.60-stabe`.
I'll add the comments on the next PR.

**Pull request merged**